src/test.js

```diff
@@ -674,63 +674,120 @@ function checkPollution() {
 	}
 }
 
+function addTestWithData( data ) {
+	if ( focused || config.currentModule.ignored ) {
+		return;
+	}
+
+	const newTest = new Test( data );
+
+	newTest.queue();
+}
+
 let focused = false; // indicates that the "only" filter was used
 
 // Will be exposed as QUnit.test
 export function test( testName, callback ) {
+	addTestWithData( {
+		testName: testName,
+		callback: callback
+	} );
+}
+
+function todo( testName, data, callback ) {
 	if ( focused || config.currentModule.ignored ) {
 		return;
 	}
 
 	const newTest = new Test( {
+		testName,
+		callback,
+		todo: true,
+		params: data
+	} );
+
+	newTest.queue();
+}
+
+function skip( testName ) {
+	if ( focused || config.currentModule.ignored ) {
+		return;
+	}
+
+	const test = new Test( {
 		testName: testName,
-		callback: callback
+		skip: true
+	} );
+
+	test.queue();
+}
+function only( testName, data, callback ) {
+	if ( config.currentModule.ignored ) {
+		return;
+	}
+	if ( !focused ) {
+		config.queue.length = 0;
+		focused = true;
+	}
+
+	const newTest = new Test( {
+		testName: testName,
+		callback: callback,
+		params: data
 	} );
 
 	newTest.queue();
 }
 
-extend( test, {
-	todo: function todo( testName, callback ) {
-		if ( focused || config.currentModule.ignored ) {
-			return;
-		}
+function runEach( data, eachFn ) {
+	if ( Array.isArray( data ) ) {
+		data.forEach( eachFn );
+	} else {
+		throw new Error(
+			`test.each expects an array of arrays or an array of primitives as the expected input.
+${typeof data} was found instead.`
+		);
+	}
+}
 
-		const newTest = new Test( {
-			testName,
-			callback,
-			todo: true
+extend( test, {
+	todo: function( testName, callback ) {
+		todo( testName, undefined, callback );
+	},
+	skip,
+	only: function( testName, callback ) {
+		only( testName, undefined, callback );
+	},
+	each: function( testName, data, callback ) {
+		runEach( data, ( datum, i ) => {
+			addTestWithData( {
+				testName: `${i} ${testName}`,
+				callback: callback,
+				params: datum
+			} );
 		} );
+	}
+} );
 
-		newTest.queue();
-	},
-	skip: function skip( testName ) {
-		if ( focused || config.currentModule.ignored ) {
-			return;
-		}
 
-		const test = new Test( {
-			testName: testName,
-			skip: true
+extend( test.each, {
+	todo: function( testName, data, callback ) {
+		runEach( data, ( datum, i ) => {
+			todo( `${i} ${testName}`, datum, callback );
 		} );
-
-		test.queue();
 	},
-	only: function only( testName, callback ) {
-		if ( config.currentModule.ignored ) {
-			return;
-		}
-		if ( !focused ) {
-			config.queue.length = 0;
-			focused = true;
-		}
-
-		const newTest = new Test( {
-			testName: testName,
-			callback: callback
+	skip: function( testName, data ) {
+		runEach( data, ( _, i ) => {
+			skip( `${i} ${testName}` );
 		} );
+	}
+} );
 
-		newTest.queue();
+extend( test.only, {
```

Comment:
I think we'll have just the one property here for now, so might do without `extend()` and assign `test.only.each` instead, perhaps?

Reply:
Makes sense, done.